File: Binary Tree/145.二叉树的后序遍历.cpp
#include <vector>

using namespace std;

/*
 * @lc app=leetcode.cn id=145 lang=cpp
 *
 * [145] 二叉树的后序遍历
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution
{
public:
    // 中序遍历遍历
    vector<int> result;
    void perTraver(TreeNode *root, vector<int> &result)
    {
        if (root == nullptr)
        {
            return;
        }

        // 遍历左子树
        perTraver(root->left, result);
        // 遍历右子树
        perTraver(root->right, result);

        // 遍历中间节点
        result.push_back(root->val);
    }
    // 前序遍历
    vector<int> postorderTraversal(TreeNode *root)
    {
        perTraver(root, result);
        return result;
    }
};
// @lc code=end

File: Binary Tree/二叉树的遍历方式/145.二叉树的后序遍历.ts
/*
 * @lc app=leetcode.cn id=145 lang=typescript
 *
 * [145] 二叉树的后序遍历
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function postorderTraversal(root: TreeNode | null): number[] {

  let result = []

  const endTraversal = (tree: TreeNode | null): void => {
    if (!tree) {
      return
    }
    endTraversal(tree.left)
    endTraversal(tree.right)
    result.push(tree.val)
  }

  endTraversal(root)

  return result
};

// @lc code=end

